Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,483 @@
+            case 2: quotient.lo = (((remainder / denominator) << 32) | quotient.lo & 0x0000_0000_FFFF_FFFF)
+            case 3: quotient.lo = (((remainder / denominator) & 0xFFFF_FFFF) | quotient.lo & 0xFFFF_FFFF_0000_0000)
+            default: _ = ()
+            }
+            /* Store the remainder */
+            remainder %= denominator
+        }
+
+        return (quotient: quotient, remainder: Int(remainder & 0xFFFF_FFFF))
+    }
+}
+
+public struct BSONDecimal128: Equatable, Hashable, CustomStringConvertible {
+    // swiftlint:disable line_length
+    private static let digitsRegex = #"(?:\d+)"#
+    private static let indicatorRegex = #"(?:e|E)"#
+    private static let signRegex = #"[+-]"#
+    private static let infinityRegex = #"Infinity|Inf|infinity|inf"#
+    private static let decimalRegex = "\(digitsRegex)\\.\(digitsRegex)?|\\.?\(digitsRegex)"
+    private static let nanRegex = #"NaN"#
+    private static let exponentRegex = "\(indicatorRegex)(\(signRegex))(\(digitsRegex))"
+    private static let numericValueRegex = "(\(signRegex))?(?:(\(decimalRegex))(?:\(exponentRegex))?|(\(infinityRegex)))"
+    public static let decimal128Regex = "\(numericValueRegex)|(\(nanRegex))"
+    // swiftlint:enable line_length
+
+    // The precision of the Decimal128 format
+    private static let significandDigits = 34
+    // NOTE: the min and max values are adjusted for when the decimal point is rounded out
+    // e.g, 1.000...*10^-6143 == 1000...*10^-6176
+    // In the spec exp_max is 6144 so we use 6111
+    private static let exponentMax = 6111
+    // In the spec exp_min is -6134 so we use -6176
+    private static let exponentMin = -6176
+    // The sum of the exponent and a constant (bias) chosen to make the biased exponent’s range non-negative.
+    private static let exponentBias = 6176
+
+    private static let decimalShift17Zeroes: UInt64 = 100_000_000_000_000_000
+    private static let exponentMask = 0x3FFF
+
+    private static let negativeInfinity = UInt128(hi: 0xF800_0000_0000_0000, lo: 0)
+    private static let infinity = UInt128(hi: 0x7800_0000_0000_0000, lo: 0)
+    private static let NaN = UInt128(hi: 0x7C00_0000_0000_0000, lo: 0)
+
+    private enum REGroups: Int, CaseIterable {
+        case sign = 1
+        case decimalPart = 2
+        case exponentSign = 3
+        case exponentPart = 4
+        case infinity = 5
+        case nan = 6
+    }
+
+    public var description: String { self.toString() }
+
+    /// Holder for raw decimal128 value
+    private var value: UInt128
+
+    /// Determines if the value is 0
+    private var isNegative: Bool { (self.value.hi >> 63) == 1 }
+
+    /// Indicators in the combination field that determine number type
+    private static let combinationNaN = 0b11111
+    private static let combinationInfinity = 0b11110
+
+    /// Determines if the value is Not a Number
+    private var isNaN: Bool { ((self.value.hi >> 58) & 0x1F) == Self.combinationNaN }
```

Comment:
Updated the comments, I switched to this method because ieee spec says the rest of NaN can be filled with junk, its all don't care values, and infinity I think is more strict but it felt right to match them up